wgpu-hal/src/dx12/command.rs

```diff
@@ -312,6 +312,18 @@ impl crate::CommandEncoder<super::Api> for super::CommandEncoder {
                 barrier.usage.start,
                 barrier.usage.end
             );
+            if barrier
+                .buffer
+                .memory_flags
+                .contains(crate::MemoryFlags::TRANSIENT)
+            {
+                unsafe {
+                    self.list
+                        .unwrap()
+                        .DiscardResource(barrier.buffer.resource.as_mut_ptr(), ptr::null());
+                }
+            }
```

Comment:
This is overzealous - we should only issue a discard if the transition is from undefined, otherwise we'll be blowing away perfectly good data.

Reply:
I was thinking that this might be ok since we only call `transition_buffers` once with the temp buffer to transition it to be the copy source but adding a check wouldn't hurt.

One thing to note is that the transition is always from `MAP_WRITE` to `COPY_SRC` (and this should be what we check for).

https://github.com/gfx-rs/wgpu/blob/110e62aed255777ac635a504ec81029483065be8/wgpu-core/src/device/queue.rs#L801-L804

Reply:
Actually, that won't work either. If we discard the resource _after_ writing the bytes to it, we'll get a potentially munged data.

Reply:
Hmm, right

Reply:
So we should add a `UNDEFINED` flag, right?

Reply:
Does this PR actually fix the issue you encountered? After rereading the docs, I'm also unsure why this would be needed.

Reply:
> Does this PR actually fix the issue you encountered? After rereading the docs, I'm also unsure why this would be needed.

Yes, at the beginning I changed the fully transparent data to rgba (0x00, 0x00, 0x00, 0x01), and the problem no longer appeared.   

Later, I searched for the commit history of `wgpu` to locate the `suballocation` feature, and then I disable `suballocation` to solve this problem, because the `CommittedResource` will be initialized automatically.   

I suspect that some hardware or drivers will assume that the buffer has been initialized with 0 value, so some write operations may be skipped, the documentation of msdn does only mention render target or depth stencil need to be initialized, but from my actual test results, `DiscardResource` can solve this problem.